ai_feedback/image_processing.py

```diff
@@ -106,27 +107,45 @@ def process_image(args, prompt: dict) -> tuple[str, str]:
 
     requests: list[str] = []
     responses: list[str] = []
+    
     for question in questions:
-        message = Message(role="user", content=prompt["prompt_content"], images=[])
-
+        # Start with the raw prompt content
+        prompt_content = prompt["prompt_content"]
+        
         # Always replace {context} when it appears
-        if "{context}" in message.content:
+        if "{context}" in prompt_content:
             context = read_question_context(OUTPUT_DIRECTORY, question)
-            message.content = message.content.replace(
+            prompt_content = prompt_content.replace(
                 "{context}", "```\n" + context + "\n```"
             )
-        if "{image_size}" in message.content:
+        if "{image_size}" in prompt_content:
             submission_image_path = args.submission_image
             # Only consider one image per question
             image = PILImage.open(submission_image_path)
-            message.content = message.content.replace(
+            prompt_content = prompt_content.replace(
                 "{image_size}", f"{image.width} by {image.height}"
             )
-        if prompt.get("include_submission_image", False):
+            
+        rendered_prompt = render_prompt_template(prompt_content)
+        
+        # Replace generic image placeholders with specific positional references
+        has_submission = "{submission_image}" in prompt_content
+        has_solution = "{solution_image}" in prompt_content and args.solution_image
+
+        if has_submission and has_solution:
```

Comment:
Okay this is definitely cleaner, but I think you can improve further by merging this logic into the `render_prompt_template` function. There isn't a good reason to do two substitutions, you can replace `{submission_image}` and `{solution_image}` directly with the final text.